### ISSUE TITLE
Debugging Not Started Says  "could not find the app launcher by the project root path" Error Code 1410

### ISSUE THREAD

**Issue by ashadbubt** [microsoft/vscode-react-native#1582]
# 🐛 Debugging Not Started Says  "could not find the app launcher by the project root path" Error Code 1410?

I Previously test and debugging work Perfectly But Now i just start my  project and  try to to start debugging it says this error .
Then I crate a new fresh react native project    and still same problem  
What I did 
1. Create new Project By Command ( create-react-native-app debugtest )
2. Install React Native Tools  v1.5.0
3. Configure Packager Info
{
  "expoServerPort": 19001,
  "packagerPort": 19001,
  "packagerPid": null,
  "devToolsPort": 19003,
  "expoServerNgrokUrl": null,
  "packagerNgrokUrl": null,
  "ngrokPid": null,
  "webpackServerPort": null
}
4. configure react native tolls 
  {
          "name": "Attach to packager",
          "cwd": "${workspaceFolder}",
          "type": "reactnative",
          "port": "19000",
          "request": "attach"
 }

## Expected behavior

It shoud start debugging 

## Debug output

<details>
<summary>Debug Console</summary>

```text

PASTE OUTPUT OF DEBUG CONSOLE (View -> Toggle Debug Console)

```
</details>

<details>
<summary>React Native output channel</summary>

```text

PASTE OUTPUT OF REACT-NATIVE OUTPUT CHANNEL (View -> Toggle Output -> Select React-Native in ListBox)

```
</details>

<details>
<summary>Developer Tools console</summary>

```text

PASTE OUTPUT OF DEVELOPER TOOLS CONSOLE (Help -> Toggle Developer Tools -> Select Console tab)

```
</details>

## Environment

Please tell us about your system and your project:

* `npx react-native doctor` or `npx expo doctor` if your project is using Expo:

```text
Common
 ✓ Node.js
 ✓ yarn
 ✖ Python
   - Version found: 3.9.1
   - Version supported: >= 2.x < 3

Android
 ✓ JDK
 ✓ Android Studio - Required for building and installing your app on Android
 ✖ Android SDK - Required for building and installing your app on Android
   - Versions found: N/A
   - Version supported: 29.0.3
 ✓ ANDROID_HOME

Errors:   2
Warnings: 0
```

* `envinfo`:

```text
 System:
    OS: Windows 10 10.0.19041
    CPU: (12) ia32 AMD Ryzen 5 3600 6-Core Processor
    Memory: 7.31 GB / 15.95 GB
  Binaries:
    Node: 12.18.4 - C:\Program Files (x86)\nodejs\node.EXE
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 6.14.6 - C:\Program Files (x86)\nodejs\npm.CMD
  Managers:
    Composer: 1.10.1 - C:\ProgramData\ComposerSetup\bin\composer.BAT
    pip2: 1.5.6 - ~\AppData\Local\python2\scripts\pip2.EXE
    pip3: 20.2.3 - C:\Python39\Scripts\pip3.EXE
  Utilities:
    Git: 2.25.1. - C:\Program Files\Git\cmd\git.EXE
  IDEs:
    Android Studio: Version  4.1.0.0 AI-201.8743.12.41.7199119
    VSCode: 1.56.2 - C:\Users\User\AppData\Local\Programs\Microsoft VS Code\bin\code.CMD
  Languages:
    Java: 16.0.1 - C:\Program Files\Common Files\Oracle\Java\javapath\javac.EXE
    PHP: 7.4.3 - C:\xampp\php\php.EXE
    Python: 3.9.1 - C:\Python39\python.EXE
  Databases:
    SQLite: 3.32.2 - C:\Users\User\AppData\Local\Android\Sdk\platform-tools\sqlite3.EXE
  Browsers:
    Chrome: 90.0.4430.93
    Edge: Spartan (44.19041.964.0), Chromium (91.0.864.37)
    Internet Explorer: 11.0.19041.1
```

* React Native Tools extension version: v1.5.0
* Expo SDK version (if applicable): VERSION

**Comment by RedMickey** [comment]
Hi @ashadbubt and thanks for reaching us.
We tried to reproduce the issue following your repro steps:
1. We created a project and configured it
1. Ran the project using `expo start` command
  (It's also possible to run Expo apps in the extension using the `Debug in Exponent` debugging scenario) 
1. Opened the project root folder in a new VS Code window
1. Attached to the Expo app on Android using the `Attach to packager` debugging scenario

While attaching we didn't face any problems.

We'd like to know if you open the project root folder in a new VS Code window?
Could you please also tell us if you changed the `react-native-tools.projectRoot` parameter in the extension settings?

You can also check if a project was recognized as a React Native one following these steps:
1. Enable `Trace` logging by adding a `settings.json` file to the `.vscode` folder in your project with the following content:
    ```
    {
        "react-native-tools.logLevel": "Trace"
    }
    ```
1. Reload VS Code
1. If a React Native project is successfully recognized, the extension will activate, and you will see something like this:
    ![image](https://user-images.githubusercontent.com/33267199/120171676-f6bd6180-c20a-11eb-8cd0-69aff8e3f384.png)
    In this case an app launcher instance should exist.

**Comment by vcellu** [comment]
Hi, I'm having the same issue.  And yes my project root is changed.  It worked fine in 1.4.2.

**Comment by ashadbubt** [comment]
Dear @RedMickey  Bro I tried again but still says same error 
After Setting 
  "react-native-tools.logLevel": "Trace"
This it show  on that tab 

[Debug] Begin to activate...

[Debug] Extension version: 1.5.0

[Debug] Projects found: 1

[Debug] Add project: c:\xampp\htdocs\debugTest\${workspaceFolder}

[Debug] React Native version: errorUnknown

[Debug] react-native package version is not found in c:\xampp\htdocs\debugTest\${workspaceFolder}. Reason: errorUnknown

**Comment by ashadbubt** [comment]
@RedMickey  Will It work 
This is output and my folder structure. 
![Untitled](https://user-images.githubusercontent.com/16360848/120296952-72daa680-c2ea-11eb-8bce-a5d01219c485.png)

**Comment by RedMickey** [comment]
Hi @vcellu and @ashadbubt, while investigating we revealed a problem with handling the `react-native-tools.projectRoot` parameter and proposed the fix for the issue in [this PR](https://github.com/microsoft/vscode-react-native/pull/1583). We will test it and release in the next version of the extension.

@ashadbubt, thanks for your answer. It seems that your `react-native-tools.projectRoot` parameter does not contain a full or relative path, so the extension cannot find a project by that path. Could you please clean `projectRoot` parameter, and in this case the extension will use the `cwd` value from the debugging scenario, or set the full path to your project in the `react-native-tools.projectRoot` parameter?

**Comment by RedMickey** [comment]
You can try out the extension with the fix before the next release by following these steps:

- Download and unzip compiled extension [VSIX](https://github.com/microsoft/vscode-react-native/files/6575566/vscode-react-native-1.5.0.vsix.zip)
- Uninstall the extension
- Close VS Code
- Open VS Code
- Click `Extensions -> ... -> Install from VSIX...` and select downloaded VSIX

**Comment by RedMickey** [comment]
Hi @ashadbubt, @vcellu, we've released React Native Tools v1.5.1 with the fix for the issue with customized project root. Please try it out and let us know if the problem persists.

**Comment by vcellu** [comment]
> Hi @ashadbubt, @vcellu, we've released React Native Tools v1.5.1 with the fix for the issue with customized project root. Please try it out and let us know if the problem persists.
> 
> 

This release works for me :)

**Comment by ashadbubt** [comment]
@RedMickey  Its work if i just set exact location of project .
if i set "React-native-tools: Project Root"  this value 
"C:\xampp\htdocs\debugTest" instead of ${workspaceFolder} . its work but when i set "${workspaceFolder} " it not working.
but previously it work with this "${workspaceFolder} " value

**Comment by RedMickey** [comment]
@ashadbubt, I tried to set `${workspaceFolder}` value to the `react-native-tools.projectRoot` parameter in React Native Tools v1.4.2 and also got an error.
![image](https://user-images.githubusercontent.com/33267199/121016505-e0804a00-c7a4-11eb-9480-e6272b5d805c.png)

In accordance with the [`Variables Reference` documentation ](https://code.visualstudio.com/docs/editor/variables-reference) predefined variables substitution is only supported in Debugging and Task configuration files. So the `react-native-tools.projectRoot` parameter in the extension settings should only support string values such as an absolute or relative path.
Since RNT v1.5.0 we've improved multi-root support and React Native version checking functional, so now while getting an RN version the extension uses an `appLauncher` instance available by the project root path. So if a project root path, specified in the `projectRoot` parameter or in the `cwd` property (in `launch.json`) is incorrect, the extension can throw an error.
If you don't use multi-root approach, it's not necessary to specify the `projectRoot` parameter, the extension will use the value of the `cwd` property from the debugging configuration.
